pyface/data_view/i_data_view_widget.py

```diff
@@ -8,52 +8,283 @@
 
-    def _get_control_header_visible(self):
-        """ Toolkit specific method to get the visibility of the header. """
-        raise NotImplementedError()
+    @contextmanager
+    def _selection_updating(self):
+        """ Context manager to prevent loopback when updating selections. """
+        if self._selection_updating_flag:
+            yield
+        else:
+            self._selection_updating_flag = True
+            try:
+                yield
+            finally:
+                self._selection_updating_flag = False
 
-    def _set_control_header_visible(self, tooltip):
-        """ Toolkit specific method to set the visibility of the header. """
-        raise NotImplementedError()
+    # Trait property handlers
+
+    @cached_property
+    def _get_selection(self):
+        return self._selection
+
+    def _set_selection(self, selection):
+        if self.selection_mode == 'none' and len(selection) != 0:
+            raise TraitError(
+                "Selection must be empty when selection_mode is 'none', "
+                "got {!r}".format(selection)
+            )
+        elif self.selection_mode == 'single' and len(selection) > 1:
+            raise TraitError(
+                "Selection must have at most one element when selection_mode "
+                "is 'single', got {!r}".format(selection)
+            )
+
+        if self.selection_type == 'row':
+            for row, column in selection:
+                if column != ():
+                    raise TraitError(
+                        "Column values must be () when selection_type is "
+                        "'row', got {!r}".format(column)
+                    )
+                if not self.data_model.is_row_valid(row):
+                    raise TraitError(
+                        "Invalid row index {!r}".format(row)
+                    )
+        elif self.selection_type == 'column':
+            for row, column in selection:
+                if not (self.data_model.is_row_valid(row)
+                        and self.data_model.can_have_children(row)
+                        and self.data_model.get_row_count(row) > 0):
+                    raise TraitError(
+                        "Row values must have children when selection_type "
+                        "is 'column', got {!r}".format(column)
+                    )
+                if not self.data_model.is_column_valid(column):
+                    raise TraitError(
+                        "Invalid column index {!r}".format(column)
+                    )
+        else:
+            for row, column in selection:
+                if not self.data_model.is_row_valid(row):
+                    raise TraitError(
+                        "Invalid row index {!r}".format(row)
+                    )
+                if not self.data_model.is_column_valid(column):
+                    raise TraitError(
+                        "Invalid column index {!r}".format(column)
+                    )
```

Comment:
Yes, this is a limitation, and it is a good use-case for the new `TraitList` etc.  But solving this in complete generality is hard, even for the limited length constraints we have.  I would like to see it in the future, if possible, but Traits isn't up to it yet.

The saving grace is that there will eventually be a traitsui editor which sits in front of this class and passes the selection list back and forth from. some named trait on a `HasTraits`.  This would be the wrong place to put that code, but it does mean that almost all modifications of the list will be via setting the complete list, rather than in-place modification.

There are other potential work-arounds:

- making the selection a private trait and adding get/set methods that _have_ to be used.
- have a selection object that wraps the selection and more tightly controls the interactions, but which can validate more closely
- use a tuple or other immutable sequence rather than a list to prevent mutation (but that restricts the validation that we can perform with Traits).

But here I'm going for the minimal thing that works.  Users of this class will simply have to take care to only modify the selection by replacement, not mutation.

Reply:
Actually, what I really want is something like traitlets `@validate` method